File: dataset/xCodeEval_4000/C++/Code/ff8b3d5c66c337d07789a8eff6fb3621.cpp
#include <cstring>
#include <cstdio>
#include <vector>
#include <algorithm>
#include <iostream>
#include <cmath>
#include <utility>
using namespace std;

#define _inline(f...) f() __attribute__((always_inline)); f
#define all(v) (v).begin(), (v).end()
#define rall(v) (v).rbegin(), (v).rend()
const int INF = 10000000;
const double EPS = 1e-8;

_inline(int cmp)(double x, double y = 0, double tol = EPS) {
  return (x <= y + tol) ? (x + tol < y) ? -1 : 0 : 1;
}

struct point {
  double x, y;
  point(double x = 0, double y = 0): x(x), y(y) {}
  
  point operator +(point q) { return point(x + q.x, y + q.y); }
  point operator -(point q) { return point(x - q.x, y - q.y); }
  point operator *(double t) { return point(x * t, y * t); }
  point operator /(double t) { return point(x / t, y / t); }
  double operator *(point q) { return x * q.x + y * q.y; }
  double operator %(point q) { return x * q.y - y * q.x; }
  int cmp(point q) const {
    if (int t = ::cmp(x, q.x)) return t;
    return ::cmp(y, q.y);
  }
  bool operator ==(point q) const { return cmp(q) == 0; }
  bool operator !=(point q) const { return cmp(q) != 0; }
  bool operator < (point q) const { return cmp(q) < 0; }
  friend ostream& operator <<(ostream& o, point p) {
    return o << "(" << p.x << ", " << p.y << ")";
  }
  
  static point pivot;
};

point point::pivot;

double abs(point p) { return hypot(p.x, p.y); }
double arg(point p) { return atan2(p.y, p.x); }

typedef vector<point> polygon;
typedef pair<point, double> circle;

_inline(int ccw)(point p, point q, point r) {
  return cmp((p-r) % (q-r));
}









_inline(double angle)(point p, point q, point r) {
  point u = p - q, v = r - q;
  return atan2(u % v, u * v);
}

double dist2(point p, point q) { return ((p-q) *  (p-q)); }


point rotateCCW90(point p) { return point(-p.y,p.x); }
point rotateCW90(point p) { return point(p.y,-p.x); }
point rotateCCW(point p, double t) { 
  return point(p.x*cos(t)-p.y*sin(t), p.x*sin(t)+p.y*cos(t)); 
}











point projectPointLine(point a, point b, point c) {
  return a + (b-a)*((c-a)* (b-a))/ ((b-a) * (b-a));
}


point projectPointSegment(point a, point b, point c) {
  double r = ((b-a) * (b-a));
  if (fabs(r) < EPS) return a;
  r = ((c-a) * (b-a))/r;
  if (r < 0) return a;
  if (r > 1) return b;
  return a + (b-a)*r;
}





bool between(point p, point q, point r) {
  return ccw(p, q, r) == 0 && cmp((p - q) * (r - q)) <= 0;
}



double distancePointSegment(point a, point b, point c) {
  return sqrt(dist2(c, projectPointSegment(a, b, c)));
}


double distancePointPlane(double x, double y, double z,
			  double a, double b, double c, double d){
  return fabs(a*x+b*y+c*z-d)/sqrt(a*a+b*b+c*c);
}


bool linesParallel(point a, point b, point c, point d) { 
  return fabs((b-a) % ( c-d)) < EPS; 
}
bool linesCollinear(point a, point b, point c, point d) { 
  return linesParallel(a, b, c, d)
    && fabs((a-b) % (a-c)) < EPS
    && fabs((c-d) % (c-a)) < EPS; 
}





bool seg_intersect(point p, point q, point r, point s) {
  point A = q - p, B = s - r, C = r - p, D = s - q;
  int a = cmp(A % C) + 2 * cmp(A % D);
  int b = cmp(B % C) + 2 * cmp(B % D);
  if (a == 3 || a == -3 || b == 3 || b == -3) return false;
  if (a || b || p == r || p == s || q == r || q == s) return true;
  int t = (p < r) + (p < s) + (q < r) + (q < s);
  return t != 0 && t != 4;
}





point line_intersect(point p, point q, point r, point s) {
  point a = q - p, b = s - r, c = point(p % q, r % s);
  return point(point(a.x, b.x) % c, point(a.y, b.y) % c) / (a % b);
}















bool circle2PtsRad(point p1, point p2, double r, point &c) {
  double d2 = (p1.x - p2.x) * (p1.x - p2.x) + (p1.y - p2.y) * (p1.y - p2.y);
  double det = ((r * r) / d2) - 0.25;
  if (det < 0.0) return false;
  double h = sqrt(det);
  c.x = (p1.x + p2.x) * 0.5 + (p1.y - p2.y) * h;
  c.y = (p1.y + p2.y) * 0.5 + (p2.x - p1.x) * h;
  return true; 
}





point computeCircleCenter(point a, point b, point c) {
  b=(a+b)/2;
  c=(a+c)/2;
  return line_intersect(b, b+rotateCW90(a-b), c, c+rotateCW90(a-c));
}







point computeCircleIncenter(point a , point b, point c){
  point d = b + rotateCCW(c-b,angle(c,b,a)/2);
  point e = c + rotateCCW(a-c,angle(a,c,b)/2);
  return line_intersect(b,d,c,e);
}





vector<point> circleLineIntersection(point a, point b, point c, double r) {
  vector<point> ret;
  b = b-a;
  a = a-c;
  double A = b *  b;
  double B = a *  b;
  double C = (a * a) - r*r;
  double D = B*B - A*C;
  if (D < -EPS) return ret;
  ret.push_back(c+a+b*(-B+sqrt(D+EPS))/A);
  if (D > EPS)
    ret.push_back(c+a+b*(-B-sqrt(D))/A);
  return ret;
}





vector<point> circleCircleIntersection(point a, point b, double r, double R) {
  vector<point> ret;
  double d = sqrt(dist2(a, b));
  if (d > r+R || d+min(r, R) < max(r, R)) return ret;
  double x = (d*d-R*R+r*r)/(2*d);
  double y = sqrt(r*r-x*x);
  point v = (b-a)/d;
  ret.push_back(a+v*x + rotateCCW90(v)*y);
  if (y > 0)
    ret.push_back(a+v*x - rotateCCW90(v)*y);
  return ret;
}





vector<  point > tangentSegments(point c, double r,  point p){
  vector< point  > resp;
  point toCenter = c - p;
  if( cmp(r - sqrt(dist2(toCenter , point(0,0)))) <= 0 ){
    resp.push_back(p);
    resp.push_back(p);
    return resp;
  }
  double angle = asin(r/sqrt(dist2(toCenter , point(0,0))));
  point uni = toCenter / sqrt(dist2(toCenter , point(0,0)));
  double tam = sqrt(dist2(toCenter , point(0,0)) - r *r);
  point origi = uni * tam;
  point p1 = p + rotateCCW(origi,angle);
  resp.push_back(p1);
  point p2 = p + rotateCCW(origi,-angle);
  resp.push_back(p2);
  return resp;
}
















int in_poly(point p, polygon& T) {
  double a = 0; int N = T.size();
  for (int i = 0; i < N; i++) {
    if (between(T[i], p, T[(i+1) % N])) return -1;
    a += angle(T[i], p, T[(i+1) % N]);
  }
  return cmp(a) != 0;
}







double poly_area(polygon& T) {
  double s = 0; int n = T.size();
  for (int i = 0; i < n; i++)
    s += T[i] % T[(i+1) % n];
  return s / 2;
}




point computeCentroid(vector<point> &p) {
  point c(0,0);
  double scale = 6.0 * poly_area(p);
  for (int i = 0; i < p.size(); i++){
    int j = (i+1) % p.size();
    c = c + (p[i]+p[j])*(p[i].x*p[j].y - p[j].x*p[i].y);
  }
  return c / scale;
}





bool radial_lt(point p, point q) {
  point P = p - point::pivot, Q = q - point::pivot;
  double R = P % Q;
  if (cmp(R)) return R > 0;
  return cmp(P * P, Q * Q) < 0;
}






polygon convex_hull(vector<point>& T) {
  int j = 0, k, n = T.size(); polygon U(n);
  point::pivot = *min_element(all(T));
  sort(all(T), radial_lt);
  for (k = n-2; k >= 0 && ccw(T[0], T[n-1], T[k]) == 0; k--);
  reverse((k+1) + all(T));
  for (int i = 0; i < n; i++) {
    

    while (j > 1 && ccw(U[j-1], U[j-2], T[i]) >= 0) j--;
    U[j++] = T[i];
  }
  U.erase(j + all(U));
  return U;
}








polygon poly_intersect(polygon& P, polygon& Q) {
  int m = Q.size(), n = P.size();
  int a = 0, b = 0, aa = 0, ba = 0, inflag = 0;
  polygon R;
  while ((aa < n || ba < m) && aa < 2*n && ba < 2*m) {
    point p1 = P[a], p2 = P[(a+1) % n], q1 = Q[b], q2 = Q[(b+1) % m];
    point A = p2 - p1, B = q2 - q1;
    int cross = cmp(A % B), ha = ccw(p2, q2, p1), hb = ccw(q2, p2, q1);
    if (cross == 0 && ccw(p1, q1, p2) == 0 && cmp(A * B) < 0) {
      if (between(p1, q1, p2)) R.push_back(q1);
      if (between(p1, q2, p2)) R.push_back(q2);
      if (between(q1, p1, q2)) R.push_back(p1);
      if (between(q1, p2, q2)) R.push_back(p2);
      if (R.size() < 2) return polygon();
      inflag = 1; break;
    } else if (cross != 0 && seg_intersect(p1, p2, q1, q2)) {
      if (inflag == 0) aa = ba = 0;
      R.push_back(line_intersect(p1, p2, q1, q2));
      inflag = (hb > 0) ? 1 : -1;
    }
    if (cross == 0 && hb < 0 && ha < 0) return R;
    bool t = cross == 0 && hb == 0 && ha == 0;
    if (t ? (inflag == 1) : (cross >= 0) ? (ha <= 0) : (hb > 0)) {
      if (inflag == -1) R.push_back(q2);
      ba++; b++; b %= m;
    } else {
      if (inflag == 1) R.push_back(p2);
      aa++; a++; a %= n;
    }
  }
  if (inflag == 0) {
    if (in_poly(P[0], Q)) return P;
    if (in_poly(Q[0], P)) return Q;
  }
  R.erase(unique(all(R)), R.end());
  if (R.size() > 1 && R.front() == R.back()) R.pop_back();
  return R;
}




const double PI = acos(-1);
int xp,yp,vp,x,y,v,r;

point origin = point(0,0);
int pode(point loc, double time){
  point nave = point(x,y);
  if(cmp( distancePointSegment(loc,nave,origin) , r ) >= 0 ){
  return cmp(time ,  abs(nave-loc)/v ) >= 0 ;
  } else {

    double distN = sqrt(dist2(nave,origin) - r * r);
    double distL = sqrt(dist2(loc,origin) - r * r);
    double angNOL = fabs(angle(nave,origin,loc)) - fabs(atan(distN/r)) - fabs(atan(distL/r));
    double menor = distN + distL + fabs(r*angNOL) ;

     return cmp(time , menor/v ) >= 0 ;
  }
}

int main(){ 
  cin >> xp >> yp >> vp >> x >> y  >> v >> r;
  int cont = 1000000;
  double R = hypot(yp, xp );
  double initial = atan2(yp, xp);
  double bot = 0;
  double timeTot = 1000000.0;
  double t1 = (2 * PI * R) / vp;
  

  double top = timeTot;
  while (top - bot > EPS) {
    double mid = (top + bot) / 2;
    

    


    

    

    double angle = ((2 * PI * mid ) / t1) + initial; 
    point loc = point( cos(angle) * R , sin(angle ) * R);
    

    if( pode(loc, mid) ){
      

      top = mid - EPS;
    } else {
      

      bot = mid + EPS;
    }
    

  }
  printf("%.7f\n",bot);
  return 0;
}

File: output/GPT4o-Mini/xCodeEval_4000/C++/Java/ff8b3d5c66c337d07789a8eff6fb3621.java
import java.util.*;
import java.io.*;

public class ff8b3d5c66c337d07789a8eff6fb3621_nc {
    static final double EPS = 1e-8;
    static final double INF = 10000000;

    static int cmp(double x, double y) {
        return (x <= y + EPS) ? (x + EPS < y) ? -1 : 0 : 1;
    }

    static class Point {
        double x, y;

        Point(double x, double y) {
            this.x = x;
            this.y = y;
        }

        Point() {
            this(0, 0);
        }

        Point add(Point q) {
            return new Point(this.x + q.x, this.y + q.y);
        }

        Point sub(Point q) {
            return new Point(this.x - q.x, this.y - q.y);
        }

        Point mul(double t) {
            return new Point(this.x * t, this.y * t);
        }

        Point div(double t) {
            return new Point(this.x / t, this.y / t);
        }

        double dot(Point q) {
            return this.x * q.x + this.y * q.y;
        }

        double cross(Point q) {
            return this.x * q.y - this.y * q.x;
        }

        int cmp(Point q) {
            if (int t = cmp(this.x, q.x)) return t;
            return cmp(this.y, q.y);
        }

        boolean equals(Point q) {
            return cmp(q) == 0;
        }

        boolean notEquals(Point q) {
            return cmp(q) != 0;
        }

        boolean lessThan(Point q) {
            return cmp(q) < 0;
        }

        public String toString() {
            return "(" + x + ", " + y + ")";
        }
    }

    static double abs(Point p) {
        return Math.hypot(p.x, p.y);
    }

    static double arg(Point p) {
        return Math.atan2(p.y, p.x);
    }

    static int ccw(Point p, Point q, Point r) {
        return cmp((p.sub(r)).cross(q.sub(r)));
    }

    static double angle(Point p, Point q, Point r) {
        Point u = p.sub(q), v = r.sub(q);
        return Math.atan2(u.cross(v), u.dot(v));
    }

    static double dist2(Point p, Point q) {
        return (p.sub(q)).dot(p.sub(q));
    }

    static Point rotateCCW90(Point p) {
        return new Point(-p.y, p.x);
    }

    static Point rotateCW90(Point p) {
        return new Point(p.y, -p.x);
    }

    static Point projectPointLine(Point a, Point b, Point c) {
        return a.add(b.sub(a).mul((c.sub(a)).dot(b.sub(a)) / (b.sub(a).dot(b.sub(a)))));
    }

    static Point projectPointSegment(Point a, Point b, Point c) {
        double r = (b.sub(a)).dot(b.sub(a));
        if (Math.abs(r) < EPS) return a;
        r = (c.sub(a)).dot(b.sub(a)) / r;
        if (r < 0) return a;
        if (r > 1) return b;
        return a.add(b.sub(a).mul(r));
    }

    static boolean between(Point p, Point q, Point r) {
        return ccw(p, q, r) == 0 && cmp((p.sub(q)).cross(r.sub(q))) <= 0;
    }

    static double distancePointSegment(Point a, Point b, Point c) {
        return Math.sqrt(dist2(c, projectPointSegment(a, b, c)));
    }

    static double distancePointPlane(double x, double y, double z, double a, double b, double c, double d) {
        return Math.abs(a * x + b * y + c * z - d) / Math.sqrt(a * a + b * b + c * c);
    }

    static boolean linesParallel(Point a, Point b, Point c, Point d) {
        return Math.abs((b.sub(a)).cross(c.sub(d))) < EPS;
    }

    static boolean linesCollinear(Point a, Point b, Point c, Point d) {
        return linesParallel(a, b, c, d) && Math.abs((a.sub(b)).cross(a.sub(c))) < EPS && Math.abs((c.sub(d)).cross(c.sub(a))) < EPS;
    }

    static boolean segIntersect(Point p, Point q, Point r, Point s) {
        Point A = q.sub(p), B = s.sub(r), C = r.sub(p), D = s.sub(q);
        int a = cmp(A.cross(C)) + 2 * cmp(A.cross(D));
        int b = cmp(B.cross(C)) + 2 * cmp(B.cross(D));
        if (a == 3 || a == -3 || b == 3 || b == -3) return false;
        if (a != 0 || b != 0 || p.equals(r) || p.equals(s) || q.equals(r) || q.equals(s)) return true;
        int t = (p.lessThan(r) ? 1 : 0) + (p.lessThan(s) ? 1 : 0) + (q.lessThan(r) ? 1 : 0) + (q.lessThan(s) ? 1 : 0);
        return t != 0 && t != 4;
    }

    static Point lineIntersect(Point p, Point q, Point r, Point s) {
        Point a = q.sub(p), b = s.sub(r), c = new Point(p.cross(q), r.cross(s));
        return new Point(new Point(a.x, b.x).cross(c), new Point(a.y, b.y).cross(c)).div(a.cross(b));
    }

    static boolean circle2PtsRad(Point p1, Point p2, double r, Point[] c) {
        double d2 = (p1.x - p2.x) * (p1.x - p2.x) + (p1.y - p2.y) * (p1.y - p2.y);
        double det = (r * r / d2) - 0.25;
        if (det < 0.0) return false;
        double h = Math.sqrt(det);
        c[0] = new Point((p1.x + p2.x) * 0.5 + (p1.y - p2.y) * h, (p1.y + p2.y) * 0.5 + (p2.x - p1.x) * h);
        return true;
    }

    static Point computeCircleCenter(Point a, Point b, Point c) {
        b = a.add(b).div(2);
        c = a.add(c).div(2);
        return lineIntersect(b, b.add(rotateCW90(a.sub(b))), c, c.add(rotateCW90(a.sub(c))));
    }

    static Point computeCircleIncenter(Point a, Point b, Point c) {
        Point d = b.add(rotateCCW(c.sub(b), angle(c, b, a) / 2));
        Point e = c.add(rotateCCW(a.sub(c), angle(a, c, b) / 2));
        return lineIntersect(b, d, c, e);
    }

    static List<Point> circleLineIntersection(Point a, Point b, Point c, double r) {
        List<Point> ret = new ArrayList<>();
        b = b.sub(a);
        a = a.sub(c);
        double A = b.dot(b);
        double B = a.dot(b);
        double C = (a.dot(a)) - r * r;
        double D = B * B - A * C;
        if (D < -EPS) return ret;
        ret.add(c.add(a.add(b.mul((-B + Math.sqrt(D + EPS)) / A)));
        if (D > EPS)
            ret.add(c.add(a.add(b.mul((-B - Math.sqrt(D)) / A)));
        return ret;
    }

    static List<Point> circleCircleIntersection(Point a, Point b, double r, double R) {
        List<Point> ret = new ArrayList<>();
        double d = Math.sqrt(dist2(a, b));
        if (d > r + R || d + Math.min(r, R) < Math.max(r, R)) return ret;
        double x = (d * d - R * R + r * r) / (2 * d);
        double y = Math.sqrt(r * r - x * x);
        Point v = (b.sub(a)).div(d);
        ret.add(a.add(v.mul(x)).add(rotateCCW90(v).mul(y)));
        if (y > 0)
            ret.add(a.add(v.mul(x)).sub(rotateCCW90(v).mul(y)));
        return ret;
    }

    static List<Point> tangentSegments(Point c, double r, Point p) {
        List<Point> resp = new ArrayList<>();
        Point toCenter = c.sub(p);
        if (cmp(r - Math.sqrt(dist2(toCenter, new Point(0, 0)))) <= 0) {
            resp.add(p);
            resp.add(p);
            return resp;
        }
        double angle = Math.asin(r / Math.sqrt(dist2(toCenter, new Point(0, 0))));
        Point uni = toCenter.div(Math.sqrt(dist2(toCenter, new Point(0, 0))));
        double tam = Math.sqrt(dist2(toCenter, new Point(0, 0)) - r * r);
        Point origi = uni.mul(tam);
        Point p1 = p.add(rotateCCW(origi, angle));
        resp.add(p1);
        Point p2 = p.add(rotateCCW(origi, -angle));
        resp.add(p2);
        return resp;
    }

    static int inPoly(Point p, List<Point> T) {
        double a = 0;
        int N = T.size();
        for (int i = 0; i < N; i++) {
            if (between(T.get(i), p, T.get((i + 1) % N))) return -1;
            a += angle(T.get(i), p, T.get((i + 1) % N));
        }
        return cmp(a) != 0 ? 1 : 0;
    }

    static double polyArea(List<Point> T) {
        double s = 0;
        int n = T.size();
        for (int i = 0; i < n; i++)
            s += T.get(i).cross(T.get((i + 1) % n));
        return s / 2;
    }

    static Point computeCentroid(List<Point> p) {
        Point c = new Point(0, 0);
        double scale = 6.0 * polyArea(p);
        for (int i = 0; i < p.size(); i++) {
            int j = (i + 1) % p.size();
            c = c.add(p.get(i).add(p.get(j)).mul(p.get(i).x * p.get(j).y - p.get(j).x * p.get(i).y));
        }
        return c.div(scale);
    }

    static boolean radialLt(Point p, Point q, Point pivot) {
        Point P = p.sub(pivot), Q = q.sub(pivot);
        double R = P.cross(Q);
        if (cmp(R) != 0) return R > 0;
        return cmp(P.dot(P), Q.dot(Q)) < 0;
    }

    static List<Point> convexHull(List<Point> T) {
        int j = 0, k, n = T.size();
        List<Point> U = new ArrayList<>(n);
        Point pivot = Collections.min(T, Comparator.comparingDouble(p -> p.y));
        Collections.sort(T, (p1, p2) -> radialLt(p1, p2, pivot) ? -1 : 1);
        for (k = n - 2; k >= 0 && ccw(T.get(0), T.get(n - 1), T.get(k)) == 0; k--);
        Collections.reverse(T.subList(k + 1, n));
        for (int i = 0; i < n; i++) {
            while (j > 1 && ccw(U.get(j - 1), U.get(j - 2), T.get(i)) >= 0) j--;
            U.add(T.get(i));
            j++;
        }
        U = U.subList(0, j);
        return U;
    }

    static List<Point> polyIntersect(List<Point> P, List<Point> Q) {
        int m = Q.size(), n = P.size();
        int a = 0, b = 0, aa = 0, ba = 0, inflag = 0;
        List<Point> R = new ArrayList<>();
        while ((aa < n || ba < m) && aa < 2 * n && ba < 2 * m) {
            Point p1 = P.get(a), p2 = P.get((a + 1) % n), q1 = Q.get(b), q2 = Q.get((b + 1) % m);
            Point A = p2.sub(p1), B = q2.sub(q1);
            int cross = cmp(A.cross(B)), ha = ccw(p2, q2, p1), hb = ccw(q2, p2, q1);
            if (cross == 0 && ccw(p1, q1, p2) == 0 && cmp(A.dot(B)) < 0) {
                if (between(p1, q1, p2)) R.add(q1);
                if (between(p1, q2, p2)) R.add(q2);
                if (between(q1, p1, q2)) R.add(p1);
                if (between(q1, p2, q2)) R.add(p2);
                if (R.size() < 2) return new ArrayList<>();
                inflag = 1;
                break;
            } else if (cross != 0 && segIntersect(p1, p2, q1, q2)) {
                if (inflag == 0) aa = ba = 0;
                R.add(lineIntersect(p1, p2, q1, q2));
                inflag = (hb > 0) ? 1 : -1;
            }
            if (cross == 0 && hb < 0 && ha < 0) return R;
            boolean t = cross == 0 && hb == 0 && ha == 0;
            if (t ? (inflag == 1) : (cross >= 0) ? (ha <= 0) : (hb > 0)) {
                if (inflag == -1) R.add(q2);
                ba++;
                b++;
                b %= m;
            } else {
                if (inflag == 1) R.add(p2);
                aa++;
                a++;
                a %= n;
            }
        }
        if (inflag == 0) {
            if (inPoly(P.get(0), Q) == 1) return P;
            if (inPoly(Q.get(0), P) == 1) return Q;
        }
        R = new ArrayList<>(new HashSet<>(R));
        if (R.size() > 1 && R.get(0).equals(R.get(R.size() - 1))) R.remove(R.size() - 1);
        return R;
    }

    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        int xp = sc.nextInt(), yp = sc.nextInt(), vp = sc.nextInt(), x = sc.nextInt(), y = sc.nextInt(), v = sc.nextInt(), r = sc.nextInt();
        int cont = 1000000;
        double R = Math.hypot(yp, xp);
        double initial = Math.atan2(yp, xp);
        double bot = 0;
        double timeTot = 1000000.0;
        double t1 = (2 * Math.PI * R) / vp;

        double top = timeTot;
        while (top - bot > EPS) {
            double mid = (top + bot) / 2;

            double angle = ((2 * Math.PI * mid) / t1) + initial;
            Point loc = new Point(Math.cos(angle) * R, Math.sin(angle) * R);

            if (pode(loc, mid, x, y, r, v)) {
                top = mid - EPS;
            } else {
                bot = mid + EPS;
            }
        }
        System.out.printf("%.7f\n", bot);
    }

    static boolean pode(Point loc, double time, int x, int y, int r, int v) {
        Point nave = new Point(x, y);
        if (cmp(distancePointSegment(loc, nave, new Point(0, 0)), r) >= 0) {
            return cmp(time, abs(nave.sub(loc)) / v) >= 0;
        } else {
            double distN = Math.sqrt(dist2(nave, new Point(0, 0)) - r * r);
            double distL = Math.sqrt(dist2(loc, new Point(0, 0)) - r * r);
            double angNOL = Math.abs(angle(nave, new Point(0, 0), loc)) - Math.abs(Math.atan(distN / r)) - Math.abs(Math.atan(distL / r));
            double menor = distN + distL + Math.abs(r * angNOL);

            return cmp(time, menor / v) >= 0;
        }
    }
}
